### ISSUE TITLE
Performance & DX improvements

### ISSUE THREAD

**Issue by avayaaggarwal** [OnePunchMonk/erasus#11]
## Context

After studying the architectural "moats" of PyTorch Lightning and Unsloth, here are high-impact improvements Erasus should adopt -- not copying their architecture, but applying their **unique ideas** to the machine unlearning domain.

---

### 1. `self.log()` in UnlearningModule

Lightning's killer DX feature: users call `self.log("forget_loss", loss)` inside any step method and the framework handles aggregation, logging backend dispatch, and distributed rank-zero semantics automatically.

**Current state:** `UnlearningModule.forget_step()` / `retain_step()` have no logging integration. Users must manually wire loggers.

**Proposed:**
```python
class MyModule(UnlearningModule):
    def forget_step(self, batch, batch_idx):
        loss = -F.cross_entropy(self.model(batch[0]), batch[1])
        self.log("forget_loss", loss)  # auto-aggregated, auto-logged
        return loss
```

### 2. `save_hyperparameters()` for self-describing results

Every `UnlearningResult` and checkpoint should embed the full configuration that produced it -- strategy name, selector, epochs, prune ratio, etc. -- so any result can be reproduced without external config files.

**Current state:** `UnlearningResult` stores loss histories and timing but not the config that produced it. Checkpoints save only `state_dict`.

### 3. Richer callback hook points

Current hooks: `on_epoch_start/end`, `on_batch_start/end`, `on_train_start/end` (6 hooks).
Lightning has 30+ hooks including `on_before_optimizer_step`, `on_after_backward`, `on_save_checkpoint`, `on_exception`.

**Priority additions for unlearning:**
- `on_before_unlearn_step` / `on_after_unlearn_step` -- inject gradient clipping, noise, etc.
- `on_coreset_selected` -- inspect/modify coreset before unlearning begins
- `on_checkpoint_save` -- embed unlearning metadata in checkpoints
- `on_exception` -- graceful recovery during long unlearning runs

### 4. Dual abstraction levels (Lightning + Fabric pattern)

Erasus has `ErasusUnlearner` (high-level) but no "Fabric" equivalent -- a minimal, opt-in toolkit where users keep their own loop but use Erasus utilities (coreset selection, metric computation, etc.) as composable primitives.

Users should be able to do:
```python
from erasus.fabric import select_coreset, compute_forgetting_quality, apply_gradient_ascent
# Use in their own training loop without ErasusUnlearner
```

---

### 5. Mixed precision (AMP) support

**Current state:** Zero mixed precision support. All strategies run in fp32.

Every strategy's `unlearn()` should support `torch.cuda.amp.autocast()`. This alone can 2x throughput and halve memory usage on modern GPUs. Should be a single flag: `ErasusUnlearner(model, precision="bf16-mixed")`.

### 6. Gradient checkpointing for unlearning

**Current state:** No gradient checkpointing.

Unlearning large models (7B+ LLMs) is impossible without gradient checkpointing. Unsloth's approach of async CPU offload for activations is ideal. Should be opt-in: `unlearner.fit(..., gradient_checkpointing=True)`.

### 7. Adaptive memory management

Unsloth auto-tunes chunk sizes based on available VRAM. Erasus could do the same for:
- Influence function computation (most memory-intensive operation)
- Fisher information estimation
- Gradient accumulation during unlearning

Instead of OOM crashes, auto-reduce batch size or chunk computation.

### 8. Fused unlearning kernels (future)

Unsloth fuses forward + loss + backward into single Triton kernels. For unlearning, the highest-impact fusion would be:
- Fused gradient-ascent-cross-entropy (avoid materializing full logits for forget loss)
- Fused influence score computation (avoid storing per-sample gradients)

This is a longer-term item but could yield 5-10x memory reduction for loss computation.

### 9. In-place gradient operations

Unsloth uses in-place ops (`add_()`, `mul_()`) everywhere to avoid temporary allocations. Audit all strategies for unnecessary tensor copies during gradient computation. Especially impactful for:
- `GradientAscentStrategy` -- can accumulate in-place
- `FisherForgettingStrategy` -- Fisher diagonal can be computed in-place
- `SCRUBStrategy` -- teacher-student gradient diff can be in-place

---

### 10. Smart defaults with escape hatches

Lightning's philosophy: "do the right thing by default, let users override everything."

Apply to Erasus:
- Auto-detect model type from architecture -> pick best strategy/selector
- Auto-enable AMP if CUDA is available and model supports it
- Auto-select batch size based on available memory
- All of the above overridable with explicit parameters

---

## Summary

| # | Improvement | Source | Impact | Effort |
|---|------------|--------|--------|--------|
| 1 | `self.log()` in UnlearningModule | Lightning | DX | Medium |
| 2 | `save_hyperparameters()` / self-describing results | Lightning | Reproducibility | Low |
| 3 | Richer callback hooks (8+ new hooks) | Lightning | Extensibility | Medium |
| 4 | Fabric-style composable primitives | Lightning | Flexibility | Medium |
| 5 | Mixed precision (AMP) support | Unsloth | 2x throughput | Medium |
| 6 | Gradient checkpointing | Unsloth | Enable large models | Medium |
| 7 | Adaptive memory management | Unsloth | Eliminate OOM | High |
| 8 | Fused unlearning kernels | Unsloth | 5-10x memory | High |
| 9 | In-place gradient operations | Unsloth | 20-40% memory | Low |
| 10 | Smart defaults with escape hatches | Both | DX | Medium |
